### PR TITLE
feat: allow computed properties in @wire if they are constants or primitives @W-14785085

### LINKED ISSUE #3910: `@wire` computed properties should throw more helpful error for non-string literals
Now that #3895 is merged, `@wire`s support computed properties, e.g.:

```js
@wire(getFoo, { [key1]: "$prop1" }) // identifier
wiredFoo;

@wire(getFoo, { ["key1"]: "$prop1" }) // string literal
wiredFoo;
```

However, technically you can put all kinds of literals into a computed property:

```js
@wire(getFoo, { [1]: "$prop1" }) // numeric literal
wiredFoo;

@wire(getFoo, { [/a/]: "$prop1" }) // regex literal
wiredFoo;

@wire(getFoo, { [null]: "$prop1" }) // null literal
wiredFoo;
```

Right now, if you do any of these things, you'll get an unhelpful Babel error message:

```
Property value expected type of string but got undefined
```

We should either provide a more helpful error message, or just support these types of literals (since they are valid JavaScript, even if they are weird since the property gets coerced to a string anyway).

### PR DESCRIPTION
**Pull request by wjhsf**
## Details

We currently allow computed properties in the `@wire` config object, in order to allow symbols as property keys. The current code, as written, assumes that a property key is an identifier or a string literal. However, that is not necessarily always the case (e.g. `{ 123: 'NumericLiteral' }`, `{ [true]: 'BooleanLiteral' }`). This PR loosens that restriction to also allow numeric, null, boolean, and bigint literals. Regular expression literals are excluded because they are not primitives, and template literals are excluded because they may contain expressions, meaning they are not guaranteed to always result in the same value (e.g. `` `${Math.random()}` ``).

Additionally, this PR restricts variables used in computed properties to only allow constants. This helps prevent accidentally using a computed value that changes, which would prevent reactivity from functioning appropriately.

Fixes #3910.

## Does this pull request introduce a breaking change?

* 🚨 Yes, it does introduce a breaking change.

## Does this pull request introduce an observable change?

* ✅ No, it does not introduce an observable change.
* ⚠️ Yes, it does include an observable change.

## GUS work item
W-14785085

### CODE REVIEW

**Comment by wjhsf** [comment]
Writing up a follow up ticket, I had a better idea that could solve this issue more robustly. https://github.com/salesforce/lwc/issues/3956#issuecomment-1901089168

File: packages/@lwc/babel-plugin-component/src/decorators/wire/validate.ts [hunk 1]
```diff
@@ -105,6 +101,60 @@ function validateWireParameters(path: NodePath, state: LwcBabelPluginPass) {
...
+
+function validateWireParameters(path: NodePath, state: LwcBabelPluginPass) {
+    const expressionArguments = path.get('expression.arguments');
+    if (Array.isArray(expressionArguments)) {
+        // Multiple arguments: should be [id, config?]
+        const [id, config] = expressionArguments;
+        validateWireId(id, path, state);
+        if (config) validateWireConfig(config, path, state);
+    } else {
+        // Single argument: should just be id
```
Comment (wjhsf) [inline packages/@lwc/babel-plugin-component/src/decorators/wire/validate.ts:155]: As far as I can tell, this never actually occurs (the code prior to this change doesn't handle this case and would have thrown). I don't actually know whether it's true that a single argument would be the ID here, but it seems a reasonable guess.

File: packages/@lwc/babel-plugin-component/src/__tests__/fixtures/wire-decorator/ignores-object-method/actual.js
Comment (wjhsf) [inline packages/@lwc/babel-plugin-component/src/__tests__/fixtures/wire-decorator/ignores-object-method/actual.js:1]: This diff is git being silly. I deleted the old file because it was a duplicate of another test, and the new file is unrelated.

File: packages/@lwc/errors/src/compiler/error-info/lwc-class.ts [hunk 2]
```diff
@@ -190,16 +190,33 @@ export const DecoratorErrors = {
...
+        code: 1199,
+        message:
+            'Cannot use a template literal as a computed property key. Instead, use a string or extract the value to a constant.',
+        level: DiagnosticLevel.Error,
+        url: '',
+    },
+
+    COMPUTED_PROPERTY_MUST_BE_CONSTANT_OR_LITERAL: {
+        code: 1200,
+        message: 'Computed property in @wire config must be a constant or primitive literal.',
```
Comment (nolanlawson) [inline packages/@lwc/errors/src/compiler/error-info/lwc-class.ts:218]:
```suggestion
        message: 'Computed property in @wire config must be a constant or primitive literal, such as a string or Symbol.',
```

Most developers probably don't know what the words "constant or primitive literal" mean – this gives a bit of a hint. 🙂

File: packages/@lwc/babel-plugin-component/src/decorators/wire/validate.ts (hunk 1 above)
Comment (nolanlawson) [reply packages/@lwc/babel-plugin-component/src/decorators/wire/validate.ts:155]: The [docs](https://developer.salesforce.com/docs/platform/lwc/guide/data-wire-service-about.html) are not clear, but it does seem that the `config` is optional. A quick search for `@wire` in SourceGraph reveals several instances of `@wire` being used with just the single argument. We should probably add a test for this.

Also you are correct, this line is not hit per code coverage.

**Comment by nolanlawson** [comment]
> 🚨 Yes, it does introduce a breaking change.

I agree, but to be clear I don't think this merits a major version bump. We only recently allowed computed properties in `@wire`, and 250 has not shipped to third-party component authors yet.
